File: CPP/NikhilSinghh/1071. Greatest Common Divisor of Strings.cpp
class Solution {
public:
    string gcdOfStrings(string str1, string str2) {
        string common;
        int n = str1.length(), m = str2.length();
        int i = 0, j = 0;
        while(i < n && j < m && str1[i] == str2[j]) {
            common.push_back(str1[i]);
            i++; j++;
        }
        i++; j++;
        common.push_back('0');
        while(i > 0) {
            i--; j--;
            common.pop_back();
            int g = i, c = 0;
            int ci = i, cj = j;
            if(g == 0)
                continue;
            while(ci < n && str1[ci] == common[c]) {
                ci++; c++;
                if(c == g)
                    c = 0;
            }
            if(ci < n || c != 0)
                continue;
            
            while(cj < m && str2[cj] == common[c]) {
                cj++; c++;
                if(c == g)
                    c = 0;
            }
            if(cj < m || c != 0)
                continue;
            
            return common;
        }
        return "";
    }
};

File: Java/Akash-212/1071_Greatest Common_Divisor_of_Strings.java
public class Solution {
    public String gcdOfStrings(String str1, String str2) {
        if (str1.equals(str2)) {
            return str1;
        }

        String result = "";
        char[] letters1 = str1.toCharArray();
        char[] letters2 = str2.toCharArray();

        for (int wordLength = 1; wordLength <= Math.min(str1.length(), str2.length()); wordLength++) {
            char[] substring = Arrays.copyOfRange(letters1, 0, wordLength);

            if (str1.length() % substring.length != 0 || str2.length() % substring.length != 0) {
                continue;
            }
            if (!chunksMatch(letters1, substring) || !chunksMatch(letters2, substring)) {
                continue;
            }

            result = new String(substring);
        }

        return result;
    }

    private boolean chunksMatch(char[] letters, char[] substring) {
        int chunkSize = substring.length;
        for (int i = 0; i < letters.length; i += chunkSize) {
            char[] chunk = Arrays.copyOfRange(letters, i, i + chunkSize);
            if (!Arrays.equals(chunk, substring)) {
                return false;
            }
        }
        return true;
    }
}
